source/frontend-docs.rst

```diff
@@ -0,0 +1,210 @@
+Front-end documentation with JSDoc
+##################################
+
+Writing maintainable and understandable front-end JavaScript code calls for writing
+particularly thorough documentation. JavaScript is an unusually flexible language in which
+protocols and expectations are rarely if ever obvious without carefully reading
+the code. It's also a language with a huge and diverse set of libraries where
+standards and conventions are in a constant state of flux. The only way to
+ensure that the developer who inherits your front-end code tomorrow or a year
+from now will understand what your program does is to carefully document your code,
+more so than you may be accustomed to doing on Python and other non-JS projects.
+
+A good way to document your JS is by using `JSDoc <http://usejsdoc.org/>`_ and its
+commenting conventions. JSDoc is an API documentation generator that can be used
+to generate a complete documentation site from comments in your JS code. The
+comment syntax associated with JSDoc, derived from that of JavaDoc, is also widely
+used in major JS projects. Whether or not you use JSDoc to generate an API
+docs site for your code, following the JSDoc conventions is a good way to ensure
+that your comments are as informative and immediately comprehensible as possible.
+
+JSDoc installation and basic setup
+==================================
+
+JSDoc can be added to an existing project as an npm dependency::
+
+    npm install jsdoc --save-dev
+
+For most projects, you will want to create a configuration file that will
+tell JSDoc such things as which files to ignore, which plugins to use, etc.
+Full instructions can be found `on the JSDoc site <http://usejsdoc.org/about-configuring-jsdoc.html>`_.
+
+Once the configuration file is set up, you can generate a JSDoc site from your
+JS by running this command::
+
+    node ./node_modules/jsdoc/jsdoc.js -c name-of-config-file.json -r path/to/js -d output/dir
+
+If your project uses ECMAScript 2015 with Babel, you need to use
+the ``jsdoc-babel`` plugin::
+
+    npm install jsdoc@3.3.x jsdoc-babel --save-dev
+
+You will then need to set up your configuration file to include ``jsdoc-babel``::
+
+    {
+      "plugins": ["node_modules/jsdoc-babel"]
+    }
+
+Assuming Babel is configured properly (i.e. that you have a valid .babelrc
+file tailored to your project's needs), JSDoc should "just work" after the
+Babel plugin is included.
+
+JSDoc is not a library and doesn't expose a full-featured API. For that
+reason, to use JSDoc from within your build process, it should be treated as
+just another command line program to be run.
+
+In Gulp, for instance, you can run the above command by means of Node's
+``child_process``, like so::
+
+    gulp.task('jsdoc', function (cb) {
+      child_exec('node ./node_modules/jsdoc/jsdoc.js -c ./name-of-config-file.json -r ./path/to/js -d ./output/dir', undefined, cb);
+    });
```

Comment:
Our documentation outlines getting all projects using gulp. so maybe assume that and don't show the raw command above, just outline how to add it to the gulpfile we expect projects have.